llvm/lib/Target/SPIRV/SPIRVGlobalRegistry.cpp

```diff
@@ -623,20 +623,29 @@ SPIRVType *SPIRVGlobalRegistry::createSPIRVType(const Type *Ty,
     return getOpTypeFunction(RetTy, ParamTypes, MIRBuilder);
   }
   if (auto PType = dyn_cast<PointerType>(Ty)) {
-    Type *ElemType = PType->getPointerElementType();
-
-    // Some OpenCL and SPIRV builtins like image2d_t are passed in as pointers,
-    // but should be treated as custom types like OpTypeImage.
-    if (auto SType = dyn_cast<StructType>(ElemType)) {
-      if (isOpenCLBuiltinType(SType))
-        return handleOpenCLBuiltin(SType, MIRBuilder, AccQual);
-      if (isSPIRVBuiltinType(SType))
-        return handleSPIRVBuiltin(SType, MIRBuilder, AccQual);
+    SPIRVType *SpvElementType;
+    // At the moment, all opaque pointers correspond to i8 element type.
+    // TODO: change the implementation once opaque pointers are supported
+    // in the SPIR-V specification. Use getNonOpaquePointerElementType()
+    // instead of getPointerElementType() once llvm is rebased to 15.
+    if (PType->isOpaque()) {
+      SpvElementType = getOrCreateSPIRVIntegerType(8, MIRBuilder);
+    } else {
+      Type *ElemType = PType->getPointerElementType();
```

Comment:
It's effectively the same since `getPointerElementType()` calls it itself, but IMO it's still preferred to use it explicitly despite the check above.

```suggestion
      Type *ElemType = PType->getNonOpaquePointerElementType();
```

Reply:
getNonOpaquePointerElementType is not present in llvm-14.  It will be fixed once we rebase to 15 (see TODO before this).
BTW I use getNonOpaquePointerElementType in the patch for the llvm-project trunk as agreed in the 5th patch review.